src/rt/sched/cown.h

```diff
@@ -967,16 +963,108 @@ namespace verona::rt
+      }
+
+      alloc->dealloc(msg->cowns, msg->count * sizeof(Cown*));
+      alloc->dealloc<sizeof(MessageBody)>(msg);
+
+      return true;
+    }
+
     /// Mute the senders participating in this message if a backpressure scan
     /// set the mutor during the behaviour. If false is returned, the caller
     /// must reschedule the senders and deallocate the senders array.
-    inline bool apply_backpressure(Cown** senders, size_t count)
+    inline bool apply_backpressure(
+      Alloc* alloc, EpochMark epoch, Cown** senders, size_t count)
     {
-      if (Scheduler::local()->mutor == nullptr)
+      auto* mutor = Scheduler::local()->mutor;
+      if (mutor == nullptr)
         return false;
 
-      Scheduler::local()->mute(senders, count);
+      // The array of senders is reused for the unmute message. Since fewer than
+      // the original count of cowns may be muted, a null terminator may be
+      // added before the end of the allocation to mark the end of the muted
+      // set.
+
       Scheduler::local()->mutor = nullptr;
+      size_t muting_count = 0;
+      for (size_t i = 0; i < count; i++)
+      {
+        auto* cown = senders[i];
+        auto bp = cown->bp_state.load(std::memory_order_relaxed);
+        const bool high_priority = bp.high_priority();
+        yield();
+        assert(bp.priority() != Priority::Low);
+
+        // The cown may only be muted if its priority is normal and its epoch
+        // mark is not `SCANNED`. Muting a scanned cown may result in the leak
+        // detector collecting the cown while it is muted.
+        if (high_priority || (cown->get_epoch_mark() == EpochMark::SCANNED))
+        {
+          cown->schedule();
+          continue;
+        }
+
+        yield();
+        Cown::acquire(cown);
+
+        if (
+#ifdef USE_SYSTEMATIC_TESTING
+          Systematic::coin(9) ||
+#endif
+          !cown->bp_state.compare_exchange_weak(
+            bp,
+            BPState() | bp.blocker() | Priority::Low | bp.has_token(),
+            std::memory_order_acq_rel))
+        {
+          assert(bp.priority() != Priority::Low);
+          yield();
+          cown->schedule();
+          Cown::release(alloc, cown);
+          continue;
+        }
+
+        Systematic::cout() << "Cown " << cown << ": backpressure state "
+                           << bp.priority() << " -> Low" << std::endl;
+        assert(!high_priority);
+        senders[muting_count++] = cown;
+        Systematic::cout() << "Mute cown " << cown << " (mutor: cown " << mutor
+                           << ")" << std::endl;
+
+        Scheduler::local()->mute_set_add(cown);
+      }
+
+      if (muting_count == 0)
+      {
+        alloc->dealloc(senders, count * sizeof(Cown*));
+        Cown::release(alloc, mutor);
+        return true;
+      }
+
+      if (muting_count < count)
+        senders[muting_count] = nullptr;
+
+      auto* body = MultiMessage::make_body(alloc, count, senders, nullptr);
```

Comment:
You're using the `nullptr` behaviour here for the mute message?  I think it might be nice to introduce an actual named behaviour, and detect it later.  It makes the code more descriptive, and less likely to hide errors, where a nullptr behaviour ends up not crashing the system.

Reply:
Done.